packages/expo/cli/start/server/DevelopmentSession.ts

```diff
@@ -0,0 +1,76 @@
+import { ExpoConfig, getConfig } from '@expo/config';
+
+import { APISettings } from '../../api/settings';
+import { updateDevelopmentSessionAsync } from '../../api/updateDevelopmentSession';
+import { getUserAsync } from '../../api/user/user';
+import * as ProjectDevices from '../project/devices';
+
+const UPDATE_FREQUENCY = 20 * 1000; // 20 seconds
+
+async function isAuthenticatedAsync() {
+  return !!(await getUserAsync().catch(() => null));
+}
+
+export class DevelopmentSession {
+  private timeout: ReturnType<typeof setTimeout>;
+
+  constructor(
+    /** Project root directory. */
+    private projectRoot: string,
+    /** Development Server URL. */
+    public url: string
+  ) {}
+
+  /**
+   * Notify the Expo servers that a project is running, this enables the Expo Go app
+   * and Dev Clients to offer a "recently in development" section for quick access.
+   *
+   * This method starts an interval that will continue to ping the servers until we stop it.
+   *
+   * @param projectRoot Project root folder, used for retrieving device installation ids.
+   * @param props.exp Partial Expo config with values that will be used in the Expo Go app.
+   * @param props.runtime which runtime the app should be opened in. `native` for dev clients, `web` for web browsers.
+   * @returns
+   */
+  public async startAsync({
+    exp = getConfig(this.projectRoot).exp,
+    runtime,
+  }: {
+    exp?: Pick<ExpoConfig, 'name' | 'description' | 'slug' | 'primaryColor'>;
+    runtime: 'native' | 'web';
+  }): Promise<void> {
+    if (APISettings.isOffline) {
+      this.stop();
+      return;
+    }
+
+    const deviceIds = await this.getDeviceInstallationIdsAsync();
+
+    if (!(await isAuthenticatedAsync()) && !deviceIds?.length) {
+      this.stop();
+      return;
+    }
+
+    await updateDevelopmentSessionAsync({
+      url: this.url,
+      runtime,
+      exp,
+      deviceIds,
+    });
+
+    this.stop();
+
+    this.timeout = setTimeout(() => this.startAsync({ exp, runtime }), UPDATE_FREQUENCY);
+  }
+
+  /** Get all recent devices for the project. */
+  private async getDeviceInstallationIdsAsync(): Promise<string[]> {
+    const { devices } = await ProjectDevices.getDevicesInfoAsync(this.projectRoot);
+    return devices?.map(({ installationId }) => installationId);
```

Comment:
This is typed as nonnull, plus the return type of this function itself is nonnull
```suggestion
    return devices.map(({ installationId }) => installationId);
```

Reply:
I've noticed the TS in this repo (or at least the CLI folder) doesn't account for null. For example:

```ts
const foo: number | null = null;
```
if you hover `foo` in `expo/cli` you get `const foo: number` if you do the same in `expo-cli` you get `const foo: null`.

Not sure if you know how to fix this.

Reply:
It is probably a tsconfig.json issue where strictNullChecks is not enabled: https://www.typescriptlang.org/tsconfig#strictNullChecks. If tsc isn't loading our tsconfig.json and is using the default rules, that could be the issue.